python/lsst/afw/image/_visitInfo.py

```diff
@@ -22,12 +22,105 @@
 """
 
 from lsst.utils.deprecated import deprecate_pybind11
+from lsst.utils import continueClass
 from ._imageLib import VisitInfo
 
 
 __all__ = []
 
 
+@continueClass
+class VisitInfo:  # noqa: F811
+    def copyWith(
+        self,
+        exposureId=None,
+        exposureTime=None,
+        darkTime=None,
+        date=None,
+        ut1=None,
+        era=None,
+        boresightRaDec=None,
+        boresightAzAlt=None,
+        boresightAirmass=None,
+        boresightRotAngle=None,
+        rotType=None,
+        observatory=None,
+        weather=None,
+        instrumentLabel=None,
+        id=None,
+        focusZ=None,
+        observationType=None,
+        scienceProgram=None,
+        observationReason=None,
+        object=None,
+        hasSimulatedContent=None,
+    ):
+        if exposureTime is None:
+            exposureTime = self.exposureTime
+        if darkTime is None:
+            darkTime = self.darkTime
+        if date is None:
+            date = self.date
+        if ut1 is None:
+            ut1 = self.ut1
+        if era is None:
+            era = self.era
+        if boresightRaDec is None:
+            boresightRaDec = self.boresightRaDec
+        if boresightAzAlt is None:
+            boresightAzAlt = self.boresightAzAlt
+        if boresightAirmass is None:
+            boresightAirmass = self.boresightAirmass
+        if boresightRotAngle is None:
+            boresightRotAngle = self.boresightRotAngle
+        if rotType is None:
+            rotType = self.rotType
+        if observatory is None:
+            observatory = self.observatory
+        if weather is None:
+            weather = self.weather
+        if instrumentLabel is None:
+            instrumentLabel = self.instrumentLabel
+        if id is None:
+            id = self.id
+        if focusZ is None:
+            focusZ = self.focusZ
+        if observationType is None:
+            observationType = self.observationType
+        if scienceProgram is None:
+            scienceProgram = self.scienceProgram
+        if observationReason is None:
+            observationReason = self.observationReason
+        if object is None:
+            object = self.object
+        if hasSimulatedContent is None:
+            hasSimulatedContent = self.hasSimulatedContent
+
+        return VisitInfo(
+            exposureId=0,
```

Comment:
Shouldn't that be `exposureId=exposureId`? And doesn't that then need an accompanying block above like all the others? Or am I missing something about that being special somehow?

Reply:
Oh, or are you saying that because it's a copy, it no longer corresponds to the original expId, and therefore must be zeroed out? If so, might be worth setting a more obviously suspicious value like `-1`, and either way, a comment to say that's why it's hard-coded to zero.

Reply:
Oh, this was me being lazy and just trying to make the tests work in time for the end of the day. The issue is that exposureId is deprecated (see https://github.com/lsst/afw/blob/main/python/lsst/afw/image/_visitInfo.py#L31-L35), so a newly generated exposureId is None. However the CPP API won't let a new `VisitInfo` be initialized with `None`, it must be zero. But I can put in a check for `self.exposureId is None` and set it to zero, that way this function will still be compatible with the deprecated API until it is removed.

Reply:
Ah, gotcha OK, thanks.